core/services/relay/evm/mercury/wsrpc/cache/cache.go

```diff
@@ -0,0 +1,404 @@
+		time.Now(), // first result is always "expired" and requires fetch
+	})
+	v := vi.(*cacheVal)
+
+	// HOT PATH
+	v.RLock()
+	if time.Now().Before(v.expiresAt) {
+		// CACHE HIT
+		promCacheHitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+
+		defer v.RUnlock()
+		return v.val, nil
+	} else if v.fetching {
+		// CACHE WAIT
+		promCacheWaitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+		// if someone else is fetching then wait for the fetch to complete
+		ch := v.fetchCh
+		v.RUnlock()
+		return v.waitForResult(ctx, ch, m.chStop)
+	}
+	// CACHE MISS
+	promCacheMissCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+	// fallthrough to cold path and fetch
+	v.RUnlock()
+
+	// COLD PATH
+	v.Lock()
+	if time.Now().Before(v.expiresAt) {
+		// CACHE HIT
+		promCacheHitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+		defer v.RUnlock()
+		return v.val, nil
+	} else if v.fetching {
+		// CACHE WAIT
+		promCacheWaitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+		// if someone else is fetching then wait for the fetch to complete
+		ch := v.fetchCh
+		v.Unlock()
+		return v.waitForResult(ctx, ch, m.chStop)
+	}
+	// CACHE MISS
+	promCacheMissCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+	// initiate the fetch and wait for result
+	ch := v.initiateFetch()
+	v.Unlock()
+
+	ok := m.IfStarted(func() {
+		m.wg.Add(1)
+		go m.fetch(req, v)
+	})
+	if !ok {
+		err := fmt.Errorf("memCache must be started, but is: %v", m.State())
+		v.abandonFetch(err)
+		return nil, err
+	}
+	return v.waitForResult(ctx, ch, m.chStop)
+}
+
+const minBackoffRetryInterval = 50 * time.Millisecond
+
+// newBackoff creates a backoff for retrying
+func (m *memCache) newBackoff() backoff.Backoff {
+	min := minBackoffRetryInterval
+	max := m.latestPriceTTL / 2
+	if min > max {
+		// avoid setting a min that is greater than max
+		min = max
+	}
+	return backoff.Backoff{
+		Min:    min,
+		Max:    max,
+		Factor: 2,
+		Jitter: true,
+	}
+}
+
+// fetch continually tries to call FetchLatestReport and write the result to v
+// it writes errors as they come up
+func (m *memCache) fetch(req *pb.LatestReportRequest, v *cacheVal) {
+	defer m.wg.Done()
+	b := m.newBackoff()
+	memcacheCtx, cancel := m.chStop.NewCtx()
+	defer cancel()
+	var t time.Time
+	var val *pb.LatestReportResponse
+	var err error
+	defer func() {
+		v.completeFetch(val, err, t.Add(m.latestPriceTTL))
+	}()
+
+	for {
+		var ctx context.Context
+		var cancel context.CancelFunc
+		t = time.Now()
+
+		if m.latestReportDeadline > 0 {
+			ctx, cancel = context.WithTimeoutCause(memcacheCtx, m.latestReportDeadline, errors.New("latest report fetch deadline exceeded"))
+		} else {
+			ctx = memcacheCtx
+			cancel = func() {}
+		}
```

Comment:
Could save some lines here, and make the helper call more idiomatic:
```suggestion
		ctx, cancel := memcacheCtx, func() {}
		t = time.Now()

		if m.latestReportDeadline > 0 {
			ctx, cancel = context.WithTimeoutCause(ctx, m.latestReportDeadline, errors.New("latest report fetch deadline exceeded"))
		}
```

Reply:
FWIW I think it would be OK to call `m.chStop.NewCtx()` each time too, instead of having an irregularly named memcacheCtx.